Remove the right amount of gold and exp when unchecking a task

brandonjreid opened the pull request:
There are a lot of code changes for this, but a good portion of it is splitting calculateDelta() into two functions.  The idea here is when a task is unchecked the gold and experience removed is different from the amount added.  When we uncheck a task, we need to change the task back to its original value (which involves approximating what that original value was), then calculate the delta off of the original value.  That delta can then be used for the exp and gold calculation.

I tested the approximation code for over 600,000 values between -47.27 and 21.27 and it always approximates the original value in 2 to 4 iterations, so this shouldn't affect anything speed wise.  I can upload the python code i wrote to test it, if you would like.

This doesn't fix any issues with finding an item on an accidental click, but at least the numbers come out the same now.

This fixes issue HabitRPG/habitrpg#964 along with countless other ones that I won't mention here :)

brandonjreid committed:
Changed calculateDelta() so when unchecking a task it removes the right amount of gold and exp

paglias commented:
@brandonjreid I know they are failing right now because we have to update a value but could you add a test for this? Great work, we got countless issues about that!

brandonjreid commented:
Yeah, I'll update any test cases and create some new ones.

brandonjreid committed:
Fixed bug when calculating the reverse delta.
Added in a case for a todo with checklists
so the gp and exp for those are reversed correctly.
Added a couple test cases for checking then unchecking tasks.

brandonjreid commented:
I fixed a couple bugs with what I wrote, and added some test cases.
I also put in some code for checklist todos, so the gp and exp are reversed correctly for those.
Now this should also fix issue HabitRPG/habitrpg#2222

Alys commented:
https://github.com/HabitRPG/habitrpg/issues/2222 was fixed some time ago but was never closed. I've closed it now. It was not talking about the small differences in earnings that you get from ticking/unticking, but large differences because checklist items were not being counted on unticking.

brandonjreid commented:
@Alys Ah, I see.  Well, I accidentally reintroduced the bug with my first commit, but fixed it again with the second.

paglias commented:
@brandonjreid on it now, will resolve merge conflicts and then merge. This absolutely deserve a badge, can you give me your UUID?

brandonjreid commented:
@paglias af58e0ad-3b96-49da-b03e-a5176a50c355

paglias commented:
@brandonjreid awarded :) thanks a lot
